rpc/server.go

```diff
@@ -252,16 +310,85 @@ func (s *Server) Close() {
 	}
 }
 
-// serveConn synchronously serves a single connection. When the
-// connection is closed, close callbacks are invoked.
-func (s *Server) serveConn(conn net.Conn, authHook func(proto.Message) error) {
-	s.ServeCodec(codec.NewServerCodec(conn, authHook))
-	s.mu.Lock()
-	if s.closeCallbacks != nil {
-		for _, cb := range s.closeCallbacks {
-			cb(conn)
+// readRequests synchronously reads a stream of requests from a
+// connection. Each request is handled in a new background goroutine;
+// when the handler finishes the response is written to the responses
+// channel. When the connection is closed (and any pending requests
+// have finished), we close the responses channel.
+func (s *Server) readRequests(codec rpc.ServerCodec, responses chan<- serverResponse) {
+	var wg sync.WaitGroup
+	defer func() {
+		wg.Wait()
+		close(responses)
+	}()
+
+	for {
+		req, meth, args, err := s.readRequest(codec)
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			return
+		} else if err != nil {
+			log.Warningf("rpc: server cannot decode request: %s", err)
+			return
+		}
+
+		if meth.handler == nil {
+			responses <- serverResponse{
+				req: req,
+				err: util.Errorf("rpc: couldn't find method: %s", req.ServiceMethod),
+			}
+			continue
+		}
+
+		wg.Add(1)
+		go func() {
+			reply, err := meth.handler(args)
+			responses <- serverResponse{
```

Comment:
looks like this could still rearrange responses and that the changes here have only made it less likely (since the goroutine is spawned in a serial fashion). Am I missing something?